source/pick-from-possibly-undefined.d.ts

```diff
@@ -0,0 +1,19 @@
+/**
+It allows to pick properties from a type that may be undefined
+
+@example:
+type User = {
+    id: number;
+    name: string;
+    email: string;
+} | undefined;
+
+type UserWithId = PickFromPossiblyUndefined<User, 'id' | 'name'>;
+
+Results in: UserWithId = { id: number, name: string } | undefined
+
+@category Object
+ **/
+export type PickFromPossiblyUndefined<Type, Props extends keyof NonNullable<Type>> = NonNullable<Type> extends object
+	? Pick<NonNullable<Type>, Props>
+	: NonNullable<Type>;
```

Comment:
Why return a nonNullable version of `Type`? I think this type should have the same behaviour as `Pick` except not object. or your type more like `SafePick` ?

Reply:
It won't work, TS will yell at you if you try to just Pick<Type, Props>. Using NonNullable allows you to pick from possibly undefined, that's the case. Otherwise, you'll see

`TS2345: Argument of type  unknown  is not assignable to parameter of type  string`

Reply:
> It won't work, TS will yell at you if you try to just Pick. Using NonNullable allows you to pick from possibly undefined, that's the case. Otherwise, you'll see
> 
> `TS2345: Argument of type  unknown  is not assignable to parameter of type  string`

sorry, I mean 'Why return `NonNullable<Type>` when `NonNullable<Type>
not extends object`?' 

Would it be better to return `never`?

Reply:
Okay. Let's check now, I've also added some validation so one wouldn't just do:

```
type UndefinedUserAccount = PickFromPossiblyUndefined<undefined, ''>;
```